procrustes/kopt.py

```diff
@@ -97,71 +96,59 @@ def kopt_heuristic_double(array_m, array_n, ref_error,
+    if p is None:
+        p = np.identity(np.shape(a)[0])
+    if q is None:
+        q = np.identity(np.shape(a)[0])
+
+    # compute 2-sided permutation error of the initial P & Q matrices
+    error = compute_error(a, b, p, q)
     # pylint: disable=too-many-nested-blocks
-    for comb_left in it.combinations(np.arange(num_row_left), r=kopt_k):
-        for comb_perm_left in it.permutations(comb_left, r=kopt_k):
-            if comb_perm_left != comb_left:
-                perm_kopt_left = deepcopy(perm_p)
-                # the right hand side permutation
-                for comb_right in it.combinations(np.arange(num_row_right), r=kopt_k):
-                    for comb_perm_right in it.permutations(comb_right, r=kopt_k):
-                        if comb_perm_right != comb_right:
-                            perm_kopt_right = deepcopy(perm_q)
-                            perm_kopt_right[comb_right, :] = perm_kopt_right[comb_perm_right, :]
-                            e_kopt_new_right = compute_error(array_n, array_m, perm_p.T,
-                                                             perm_kopt_right)
-                            if e_kopt_new_right < kopt_error:
-                                perm_q = perm_kopt_right
-                                kopt_error = e_kopt_new_right
-                                if kopt_error <= kopt_tol:
-                                    break
-
-                perm_kopt_left[comb_left, :] = perm_kopt_left[comb_perm_left, :]
-                e_kopt_new_left = compute_error(array_n, array_m, perm_kopt_left.T, perm_q)
-                if e_kopt_new_left < kopt_error:
-                    perm_p = perm_kopt_left
-                    kopt_error = e_kopt_new_left
-                    if kopt_error <= kopt_tol:
-                        break
-
-    return perm_p, perm_q, kopt_error
+    num_row_left = p.shape[0]
+    num_row_right = q.shape[0]
+
+    for comb_p in it.combinations(np.arange(num_row_left), r=k):
+        for perm_p in it.permutations(comb_p, r=k):
+            for comb_q in it.combinations(np.arange(num_row_right), r=k):
+                for perm_q in it.permutations(comb_q, r=k):
+                    # permute rows of matrix P
+                    p_new = deepcopy(p)
+                    p_new[comb_p, :] = p_new[perm_p, :]
+                    # permute rows of matrix Q
+                    q_new = deepcopy(q)
+                    q_new[comb_q, :] = q_new[perm_q, :]
+                    # compute error with new matrices & compare
+                    error_new = compute_error(b, a, p_new, q_new)
+                    if error_new < error:
+                        p, q, error = p_new, q_new, error_new
+                        # check whether perfect permutation matrix is found
+                        # TODO: smarter threshold based on norm of matrix
+                        if error <= 1.0e-8:
+                            break
+    return p, q, error
```

Comment:
Does this sound to you @PaulWAyers & @fwmeng88?
@PaulWAyers I have in mind that your earlier comment regarding the `1.0e-8` and double for loop applies here.

Reply:
My comment on  not explicitly constructing permutation matrices also applies here.

The rest of this looks good.

Reply:
It's a little curious that we do a row-permutation for both p_new and q_new. Since the error is |Q*A*P - B| (didn't we decide to use P1 and P2?) one could argue that row-permutations of P should be paired with column-permutations of Q.

This isn't a big-deal because a k-fold row-permutation of Q is also a k-fold column permutation of Q, so looping over all k-fold row-permutations of Q is equivalent to looping over all k-fold column-permutations of Q. So it doesn't really *matter* what's being done, but it's more intuitive (to me) to use column permutations of Q.

Reply:
I was just adding mathematical formulas to the docstrings and making the argument names consistent. I have changes the paper to use P1 and P2 and will use the same in the documentation.